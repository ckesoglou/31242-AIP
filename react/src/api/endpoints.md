// add endpoint urls here so we don't hardcode in the components + change this file format to tsx!