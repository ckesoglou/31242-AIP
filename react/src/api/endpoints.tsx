export const baseUrl = "/api";

export const loginEndpoint = baseUrl.concat("/login");
export const signUpEndpoint = baseUrl.concat("/signup");
export const userProfileEndpoint = baseUrl.concat("/user");
export const requestsEndpoint = baseUrl.concat("/requests");
export const requestsAllEndpoint = baseUrl.concat("/requests/all");
export const requestsNewEndpoint = baseUrl.concat("/requests/new");
export const imageEndpoint = baseUrl.concat("/image/");
export const itemEndpoint = baseUrl.concat("/items/");
export const leaderboardAllEndpoint = baseUrl.concat("/leaderboard/");
export const leaderboardMeEndpoint = baseUrl.concat("/leaderboard/me");
