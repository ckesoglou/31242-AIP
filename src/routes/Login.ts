import { Request, Response, Router } from "express";
import { BAD_REQUEST, OK, UNAUTHORIZED } from "http-status-codes";
import { getUser } from "../daos/Users";
import Joi, { string, ObjectSchema } from "joi";
import bcrypt from "bcrypt";
import jwt from "jsonwebtoken";
import { v4 as uuid } from "uuid";
import env from "../Environment";
import User from "../entities/User";
import Token from "@entities/Token";
import { createToken } from "@daos/Tokens";

const router = Router();

/**
 * POST: /login
 */

interface ILoginPOST {
  username: string;
  password: string;
}

const LoginPOST: ObjectSchema<ILoginPOST> = Joi.object({
  username: Joi.string().alphanum().min(2).max(16, "utf8"),
  password: Joi.string().min(1).max(72, "utf8"),
}).options({ presence: "required" });

router.post("/", async (req: Request, res: Response) => {
  const { error, value } = LoginPOST.validate(req.body);

  if (error) {
    return res.status(BAD_REQUEST).json({
      error: error.message,
    });
  }

  let requestBody = value as ILoginPOST;
  const user = await getUser(requestBody.username);

  if (user) {
    const match = await bcrypt.compare(
      requestBody.password,
      user.password_hash
    );

    if (match) {
      const now = new Date();
      const monthFromNow = new Date();
      monthFromNow.setDate(monthFromNow.getDate() + 30);

      const serverRefreshToken = await createToken(
        uuid(),
        user.username,
        req.headers.host ?? "Unknown",
        now,
        monthFromNow
      );
      const clientRefreshToken = jwt.sign(
        {
          username: user.username,
          token: serverRefreshToken.refresh_token,
        },
        env.jwt_secret,
        { expiresIn: "30 days" }
      );
      const clientAccessToken = jwt.sign(
        { username: user.username },
        env.jwt_secret,
        { expiresIn: "30m" }
      );

      res.cookie("access_tokens", {
        access_token: clientAccessToken,
        refresh_token: clientRefreshToken,
      });

      return res.status(OK).end();
    }
  }

  return res.status(UNAUTHORIZED).json({
    error: "Username or password was incorrect", // We do not specify which for security reasons
  });
});

export default router;
