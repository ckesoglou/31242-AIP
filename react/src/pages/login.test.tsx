import React from "react";
import { shallow, mount } from "enzyme";
import { Server, Response } from "miragejs";
import { baseUrl } from "../api/endpoints";

import Login from "./login";

let server: Server;

// I'm thinking that we don't
// need the server to start/stop on each test

beforeEach(() => {
  server = new Server({
    routes() {
      //   this.namespace = `${baseUrl}`;
      this.namespace = "api";

      this.post("/login", () => {
        let headers = {}; // insert new cookie here
        return new Response(200, headers);
      });
    },
  });
});

afterEach(() => {
  server.shutdown();
});

describe("<Login />", () => {
  it("should render correctly", () => {
    const wrapper = shallow(<Login />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should get a valid cookie on successful login", () => {
    const wrapper = mount(<Login />);
    let button = wrapper.find("button");
    expect(button).toHaveLength(1);
    button.simulate("click");

    // check ui changed
  });
});
