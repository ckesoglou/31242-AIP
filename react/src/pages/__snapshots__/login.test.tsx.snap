// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> should render correctly 1`] = `
<WithStyles(ForwardRef(Container))
  component="main"
  maxWidth="xs"
>
  <WithStyles(ForwardRef(Typography))
    component="h1"
    variant="h5"
  >
    Sign In
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(FormControl))>
    <WithStyles(ForwardRef(TextField))
      fullWidth={true}
      id="username"
      label="User Name"
      margin="normal"
      name="username"
      onChange={[Function]}
      required={true}
      value=""
      variant="outlined"
    />
    <WithStyles(ForwardRef(TextField))
      fullWidth={true}
      id="password"
      label="Password"
      margin="normal"
      name="password"
      onChange={[Function]}
      required={true}
      type="password"
      value=""
      variant="outlined"
    />
    <WithStyles(ForwardRef(Button))
      color="primary"
      fullWidth={true}
      onClick={[Function]}
      type="submit"
      variant="contained"
    >
      Sign In
    </WithStyles(ForwardRef(Button))>
    <WithStyles(ForwardRef(Grid))
      container={true}
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
        xs={true}
      >
        <WithStyles(ForwardRef(Link))
          href="#"
        >
          Forgot password?
        </WithStyles(ForwardRef(Link))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        item={true}
      >
        <WithStyles(ForwardRef(Link))
          href="#"
        >
          Don't have an account? Sign Up
        </WithStyles(ForwardRef(Link))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
    <div
      style={
        Object {
          "textOverflow": "ellipsis",
          "width": "10rem",
        }
      }
    >
      The token is.....
    </div>
  </WithStyles(ForwardRef(FormControl))>
</WithStyles(ForwardRef(Container))>
`;
